ouroboros-consensus/src/Ouroboros/Consensus/ChainSyncClient.hs

```diff
@@ -422,8 +428,11 @@ chainSyncClient tracer cfg btime (ClockSkew maxSkew)
 
           -- TODO: Chain sync Client: Reuse anachronistic ledger view? #581
           case anachronisticProtocolLedgerView cfg curLedger (pointSlot hdrPoint) of
-            Nothing   -> retry
-            Just view -> case view `SB.at` hdrSlot of
+            -- unexpected alternative; see comment before this case expression
+            Left TooFarBehind ->
```

Comment:
I included a specific test case which does fail if I revert the commit that added this exception, so the tip of this branch sometimes raises this exception, under circumstances involving late joins.

I didn't have full clarity when I first recognized the need for it and moreover much has changed since then.. I'll dig in and reply back here with more details on how exactly the exception is happening.

Reply:
Nick and I had a long call looking into this. We decided that `InvalidRollForward` shouldn't be necessary, but that the fault likely does not lie in this PR -- the new tests are merely exercising it.

We're going to merge as-is, and open a ticket to continue exploring why `InvalidRollForward` is catching a problematic scenario whereas we instead expect `ForkTooDeep` to have caught it before any "roll forwards" are even happening.